File: LCAofBinarySearchTree.py
# Time Complexity :
# Space Complexity :
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
# Using recursion if the p is on left side and q is on the right side of root then the LCA is root
# If both p and q are on left side then LCA will be on left and similary if both are on right side then the LCA will be on right

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode',
                             q: 'TreeNode') -> 'TreeNode':
        if not root:
            return None
        if p.val < root.val and q.val < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        elif p.val > root.val and q.val > root.val:
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return root

File: LCAOfBinarySearchTree.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace BinaryTree
{
    public class LCAOfBinarySearchTree
    {
        /*
         * T.C: O(h) where h is hight of tree
         * S.c: O(h)
         */
        public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q)
        {

            while (true)
            {
                if (root.val < p.val && root.val < q.val)
                {
                    root = root.right;
                }
                else if (root.val > p.val && root.val > q.val)
                {
                    root = root.left;
                }
                else
                {
                    return root;
                }
            }

        }

        public TreeNode lowestCommonAncestor1(TreeNode root, TreeNode p, TreeNode q)
        {

            if (root.val > p.val && root.val > q.val)
            {
                return lowestCommonAncestor1(root.left, p, q);
            }
            else if (root.val < p.val && root.val < q.val)
            {
                return lowestCommonAncestor1(root.right, p, q);
            }
            else
            {
                return root;
            }


        }
    }
}
